### PR TITLE
refactor: DRY up and support descriptors

### PR DESCRIPTION
**Pull request by buschtoens**
As a first step towards #99, this PR removes code duplication and special handling. 

I also found that Sinon uses property descriptors to stub methods in the following segment:

https://github.com/kaliber5/ember-window-mock/blob/fa60f4119a36a39cbc8c7ce4fcdfea9249d5e913/tests/unit/window-mock-test.js#L156-L179

Property descriptors were not supported by the previous proxy implementation, so I've added that as well. It only worked because the proxy stubbornly kept returning the `noop`, which made Sinon fallback to a simple assignment. But since `alert` et al aren't special handled any more, they are deleted by Sinon and the fallback doesn't work any more.

I also removed the `_reset` method. Instead you can pass the proxy to `reset`, like so:

```js
reset(proxy.location);
```

Calling `reset()` without any arguments is equivalent to calling

```js
reset(window);
```

When a proxy is reset, all properties and descriptors are discarded, including the ad-hoc proxies.

There's now also a `getProxyConfig` function that we can use to implement something like `mock` / `getHolder`.

### CODE REVIEW

File: addon-test-support/-private/setup-window-mock.js [hunk 1]
```diff
@@ -7,5 +7,5 @@ import { reset } from 'ember-window-mock';
 //
 export default function setupWindowMock(hooks = self) {
   // @TODO: assert that window mock is pristine in `hooks.before`.
-  hooks.afterEach(reset);
+  hooks.afterEach(() => reset());
```
Comment (buschtoens) [inline addon-test-support/-private/setup-window-mock.js]: Since `reset` now accepts an optional proxy, it needs to be invoked like this.

This is also why I needed to remove `tests/unit/setup-window-mock-test.js`. But since we test that `setupWindowMock` does indeed reset the mock in `tests/integration/window-test.js`, I think we're in the clear.

File: addon-test-support/-private/mock/proxy.js [hunk 2]
```diff
@@ -1,39 +1,111 @@
...
+    `Canot override non-configurable descriptor for '${key}' on '${target}'.`,
+    !descriptor || descriptor.configurable
+  );
+}
+
+function assertWritableDescriptor(target, key) {
+  const descriptor = Object.getOwnPropertyDescriptor(target, key);
+  if (descriptor && 'configurable' in descriptor && !descriptor.configurable) {
+    assert(
+      `Cannot directly set '${key}' on '${target}', because it is a non-configurable and non-writable keyerty. Instead use 'getHolder(${target}).${key}'.`,
```
Comment (buschtoens) [inline addon-test-support/-private/mock/proxy.js]: Whoops. I was assuming a specific implementation here already.

File: addon-test-support/-private/mock/proxy.js [hunk 3]
```diff
@@ -1,39 +1,111 @@
...
+    !descriptor || descriptor.configurable
+  );
+}
+
+function assertWritableDescriptor(target, key) {
+  const descriptor = Object.getOwnPropertyDescriptor(target, key);
+  if (descriptor && 'configurable' in descriptor && !descriptor.configurable) {
+    assert(
+      `Cannot directly set '${key}' on '${target}', because it is a non-configurable and non-writable property.\nSee https://github.com/kaliber5/ember-window-mock/issues/99`,
+      'set' in descriptor || descriptor.writable
```
Comment (buschtoens) [inline addon-test-support/-private/mock/proxy.js]:
```suggestion
      typeof descriptor === 'function' || descriptor.writable
```

File: addon-test-support/-private/mock/proxy.js [hunk 4]
```diff
@@ -1,39 +1,111 @@
...
+
+function getProxyConfig(proxy) {
+  assert(`Could not find '${proxy}' in the proxy map.`, PROXIES.has(proxy));
+  return PROXIES.get(proxy);
+}
+
+function assertConfigurableDescriptor(target, key) {
+  const descriptor = Object.getOwnPropertyDescriptor(target, key);
+  assert(
+    `Canot override non-configurable descriptor for '${key}' on '${target}'.`,
```
Comment (buschtoens) [inline addon-test-support/-private/mock/proxy.js]:
```suggestion
    `Cannot override non-configurable descriptor for '${key}' on '${target}'.`,
```

File: addon-test-support/-private/mock/proxy.js [hunk 5]
```diff
@@ -1,39 +1,113 @@
...
+    assert(
+      `Cannot directly set '${key}' on '${target}', because it is a non-configurable and non-writable property.\nSee https://github.com/kaliber5/ember-window-mock/issues/99`,
+      typeof descriptor.set === 'function' || descriptor.writable
+    );
+  }
+}
+
+export default function proxyFactory(
+  original,
+  { makeHolder = () => ({}), makeDescriptors = () => ({}) } = {}
```
Comment (simonihmig) [inline addon-test-support/-private/mock/proxy.js]: Uh, tbh I just learned that you can have default values in a destructuring expression, somehow didn't encounter this before. Nice! :)

File: addon-test-support/-private/setup-window-mock.js (hunk 1 above)
Comment (buschtoens) [reply addon-test-support/-private/setup-window-mock.js]: `afterEach` and all other hooks receive `assert` as an argument. `reset` takes an optional argument, a proxy, to only reset that proxy.

This means you could call `reset(window)` which is equivalent to `reset()`, but also `reset(window.navigator)` to _only_ reset the `window.navigator` proxy. This features is not strictly required, and if you prefer I can remove it. 🙃

File: addon-test-support/-private/mock/proxy.js [hunk 6]
```diff
@@ -1,39 +1,113 @@
...
+        if (config.descriptors[key].set) {
+          config.descriptors[key].set.call(original, value);
+          return true;
+        }
+        if (!config.descriptors[key].writable) {
+          return false;
+        }
+        config.descriptors[key].value = value;
+        return true;
+      }
```
[the remark this answers is not in the document]
Comment (buschtoens) [reply addon-test-support/-private/mock/proxy.js]: _Ooooooh_, that is a really neat idea! I'll try it out. 😍

File: addon-test-support/-private/setup-window-mock.js (hunk 1 above)
Comment (simonihmig) [reply addon-test-support/-private/setup-window-mock.js]: > afterEach and all other hooks receive assert as an argument. 

Ah, of course! 🤦‍♂️
I got too used to mocha, which we usually use :wink:

> This features is not strictly required, and if you prefer I can remove it. 

No, I think it adds value, so let's keep it. Hopefully `setupWindowMock()` is primarily used, so that it is not a big problem.

File: addon-test-support/-private/mock/proxy.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (buschtoens) [reply addon-test-support/-private/mock/proxy.js]: Works perfectly! 👍

File: addon-test-support/-private/mock/proxy.js [hunk 7]
```diff
@@ -1,39 +1,113 @@
...
+      assertWritableDescriptor(target, key);
+      delete config.holder[key];
+      delete target[key];
+      return true;
     },
-    deleteProperty(target, prop) {
-      delete holder[prop];
-      delete target[prop];
+    defineProperty(target, key, descriptor) {
+      assertConfigurableDescriptor(target, key);
```
[the remark this answers is not in the document]
Comment (buschtoens) [reply addon-test-support/-private/mock/proxy.js]: >> Because users would not be able to use Object.defineProperty for document in the real world, we must prohibit it here as well.
>
> Not sure why?

Turns out, you can still modify a non-configurable descriptor. 😂 
You can flip the `writable` flag for instance. But you can't switch from `value` to `get` / `set`. It's probably safest, if we copy the descriptor from `target` over to `holder` and then try mutating it. If this throws an error, we know it's not allowed.

```js
const obj = {};

Object.defineProperty(obj, 'foo', { configurable: false, writable: true, value: 'bar' });

// works
Object.defineProperty(obj, 'foo', { configurable: false, writable: false, value: 'quux' });

// breaks
Object.defineProperty(obj, 'foo', { configurable: false, get: () => 'baz', set: undefined });
```

[the remark this answers is not in the document]
Comment (buschtoens) [reply addon-test-support/-private/mock/proxy.js]: Ah, but this would then "poison" the descriptor on the `holder` as well.

[the remark this answers is not in the document]
Comment (buschtoens) [reply addon-test-support/-private/mock/proxy.js]: It's now checked using a dummy object.

[the remark this answers is not in the document]
Comment (simonihmig) [reply addon-test-support/-private/mock/proxy.js]: First, sorry for letting this wait for so long! Was a tough week... 

I am still not really understanding why we cannot just define the property on the holder? In what case would that throw an error? 

Your examples above are referring to defining a property an a regular object, that already has a descriptor. But do those examples apply also when we define the property on a holder of a Proxy, instead of the target itself?

File: addon-test-support/-private/mock/proxy.js [hunk 8]
```diff
@@ -65,8 +57,19 @@ export default function proxyFactory(original, makeHolder = () => ({})) {
...
     defineProperty(target, key, descriptor) {
-      assertConfigurableDescriptor(target, key);
+      if (!(key in config.holder)) {
+        // First copy the original descriptor to a dummy object.
+        const originalDescriptor = Object.getOwnPropertyDescriptor(target, key);
+        const dummy = {};
+        Object.defineProperty(dummy, key, originalDescriptor);
+
+        // Then try overriding it with the new descriptor, which might throw a TypeError.
+        Object.defineProperty(config.holder, key, descriptor);
```
Comment (simonihmig) [inline addon-test-support/-private/mock/proxy.js]: This is exactly the same code as 4 lines below, so what purpose does that have? And how does that help when it would throw a TypeError, as we don't catch anything?

**Comment by simonihmig** [comment]
@buschtoens could you also please add some tests covering the new support for descriptors? Especially for eventual edge cases (overriding non-writable and/or non-configurable props), where I still struggle to get a grasp of (sorry about that! :wink:), see the other comments.

File: addon-test-support/-private/mock/proxy.js (hunk 8 above)
Comment (buschtoens) [reply addon-test-support/-private/mock/proxy.js]: Right! It should be `dummy` instead of `config.holder`.

What this is supposed to do is check whether the original descriptor of the `target` is actually overridable. If it is not overridable, this line will throw an error. I did not explicitly catch that error as throwing an error in or returning `false` from `Proxy#defineProperty` will lead to the same result basically. However, to make it more explicit and to not rely on behavior which is not explicitly defined, I'll wrap this in a `try / catch` and return `false`.

The reason that we do this in the first place is that `defineProperty` will be executed, even if one of the [invariant scenarios](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/handler/defineProperty#Invariants) is hit. It _has_ to return `false` though. If it does not, a `TypeError` will be raised.

L71 `Object.defineProperty(config.holder, key, descriptor);` would have been executed nevertheless though, which means that the `defineProperty` did succeed anyway. I will add some tests around this.

Comment (buschtoens) [reply addon-test-support/-private/mock/proxy.js]: Hm, while writing the tests, I noticed some more inconsistencies. I guess I'll have to manually codify all invariants and some more rules. 🙈
